aws-lc-rs/src/agreement.rs

```diff
@@ -530,25 +575,90 @@ where
         return Err(error_value);
     }
     let peer_pub_bytes = peer_public_key.bytes.as_ref();
-    if peer_pub_bytes.len() != expected_pub_key_len {
-        return Err(error_value);
+    if peer_public_key.encoding == &OCTET_STRING {
+        let peer_pub_bytes_len = peer_pub_bytes.len();
+        if peer_pub_bytes_len != expected_pub_key_len {
+            return Err(error_value);
+        }
     }
 
     let mut buffer = [0u8; MAX_AGREEMENT_SECRET_LEN];
 
     let secret: &[u8] = match &my_private_key.inner_key {
-        KeyInner::X25519(priv_key) => {
-            x25519_diffie_hellman(&mut buffer, priv_key, peer_pub_bytes).or(Err(error_value))?
-        }
-        KeyInner::ECDH_P256(priv_key)
-        | KeyInner::ECDH_P384(priv_key)
-        | KeyInner::ECDH_P521(priv_key) => {
-            ec_key_ecdh(&mut buffer, priv_key, peer_pub_bytes, expected_nid).or(Err(error_value))?
+        KeyInner::X25519(priv_key, ..) => match peer_public_key.encoding.id {
+            EncodingID::OctetString => {
+                x25519_dh_with_octstr_peer_pubkey(
+                    peer_pub_bytes,
+                    &mut buffer,
+                    priv_key,
+                    error_value,
+                )?;
+                &buffer[0..X25519_SHARED_KEY_LEN]
+            }
+            EncodingID::X509 => {
+                x25519_dh_with_x509_peer_pubkey(
+                    peer_pub_bytes,
+                    &mut buffer,
+                    priv_key,
+                    error_value,
+                )?;
+                &buffer[0..X25519_SHARED_KEY_LEN]
+            }
+        },
+        KeyInner::ECDH_P256(ec_key) | KeyInner::ECDH_P384(ec_key) | KeyInner::ECDH_P521(ec_key) => {
+            let pub_key_bytes = peer_public_key.bytes.as_ref();
+            ec_key_ecdh(
+                &mut buffer,
+                ec_key,
+                pub_key_bytes,
+                peer_public_key.encoding,
+                expected_nid,
+            )
+            .or(Err(error_value))?
         }
     };
     kdf(secret)
 }
 
+#[inline]
+fn x25519_dh_with_octstr_peer_pubkey<E>(
+    peer_pub_bytes: &[u8],
+    shared_secret: &mut [u8; MAX_AGREEMENT_SECRET_LEN],
+    priv_key: &LcPtr<EVP_PKEY>,
+    error_value: E,
+) -> Result<(), E> {
+    let mut pub_key = [0u8; X25519_PUBLIC_VALUE_LEN as usize];
+    pub_key.copy_from_slice(peer_pub_bytes);
```

Comment:
nit: this will panic if the lengths of the two slices don't match. It might be worth requiring the caller to pass the array form (&[u8; X25519_PUBLIC_VALUE_LEN]) and/or using `.try_into().map_err(|_| error_value)?` here?

I think in practice this should be fine since have a check for the octet string length above, but it seems a bit separated out. Moving that check into this function could also work.

Reply:
Thanks.  Turns out I can get rid of `x25519_dh_with_octstr_peer_pubkey` entirely.